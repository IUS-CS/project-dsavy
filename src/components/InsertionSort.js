import React, { useState, useEffect, useRef } from 'react'
import { Container, Row, Col, Button } from 'react-bootstrap';
import Enqueue from './Forms/Enqueue'
import Dequeue from './Forms/Dequeue'

const InsertionSort = () => {
  const [array, setArray] = useState([])
  const [index, setIndex] = useState(0)
  const [running, setRunning] = useState(false)
  const [i, setI] = useState(1)
  const [delay, setDelay] = useState(50)
  const [key, setKey] = useState()
  
  const arrayStyle = {
    position: 'absolute',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    background: "linear-gradient(#58C2ED, #1B85DC)",
    width: '5rem',
    height: '18rem',
    bottom: '-2.5px'
  }
  
  const sort = (a) => {
    let temp = [...a]
    let tempIndex = index
    let tempKey = key
    
    // display each swap when comparing to key  
    if (tempIndex >= 0 && temp[tempIndex].element > tempKey) {
      temp[tempIndex + 1].element = temp[tempIndex].element
      tempIndex--
      setIndex(tempIndex)
      return temp
    }
    // now change keys and display sorted array up to this point

    setI(i + 1)
    console.log(i)

    // if array is not completely sorted set new key
    if (i < array.length) setKey(temp[i].element)

    setIndex(i - 1)
    temp[tempIndex + 1].element = tempKey
    return temp
  }

  const useInterval = (callbackfn, delay) => {
    const savedCallback = useRef();
  
    // remember the most recent callback
    useEffect(() => {
      savedCallback.current = callbackfn;
    }, [callbackfn]);
  
    // set up the interval
    useEffect(() => {
      function call() {
        savedCallback.current();
      }
      if (delay !== null) {
        let interval = setInterval(call, delay);
        return () => clearInterval(interval);
      }
    }, [delay])
  }

  useInterval(() => {
    if (running) {
        if (i < array.length+1) {
          setArray(sort(array))
        }
    }
  }, (delay !== null)? (delay*25): null)

  const handleVis = (event) => {
    event.preventDefault()
    setRunning(true)
  }

  // updates state of array 
  const onAdd = (entry) => {
    setArray(add(array, entry))
  }

  // returns array with number added to the end
  const add = (array, number) => {
    let a = [...array]
    if (number == '') {
      alert('Enter number')
      return array
    }
    
    if (array.length > 9) {
      alert('Queue is at max length')
      return array
    }

    if (Math.abs(number) > 999) {
      alert('Number must be 3 digits or less')
      return array
    }
    
    a.push({element: Number(number), search: false})  
    return a
  }

  // updates state of array
  const onDelete = () => {
    setArray(remove(array))
  }

  // returns array with first element deleted
  const remove = (array) => {
    return array.splice(1)  
  }

  const displayArray = () => {
    return (array.map((item) => 
        <li style={{listStyle: 'none'}}>
          <Col style={Object.assign({border: 'solid', borderLeft: 0, left: `${array.indexOf(item)*5}rem`}, arrayStyle)} >
            <h4 style={item.search ? {color: 'red'} : {color: 'white'}}>{item.element}</h4>
          </Col>
        </li>
    ))
  }

  return (
    <Container>
      <Row><Enqueue enqueue={onAdd} title={'Insert'}/></Row>
      <Row><Dequeue dequeue={onDelete} title={'Remove'}/></Row>
      <Row>
        <Button className="my-1" onClick={handleVis}>Sort</Button>
        <div className="my-3" style={{paddingLeft: 10}}>
          <label htmlFor="speed-range" style={{margin: '0 auto'}}>Speed</label>
          <input type="range" className="custom-range" id="speed-range" onChange={event => setDelay(event.target.value)} style={{transform: 'rotate(180deg)'}} />
        </div>
      </Row>
      <Row>
        <div style={{position: 'relative', width: '50rem', height: '18rem', margin: '0 auto' , border: 'solid'}}>
          <ul>{displayArray()}</ul>
        </div>
      </Row>
      <Row>Key being sorted: {key}</Row>
      <Row>From index {index+1}</Row>
    </Container>
  )
}

export default InsertionSort

 