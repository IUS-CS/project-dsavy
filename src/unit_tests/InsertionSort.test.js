import React from 'react';
import InsertionSort from 'components/InsertionSort';
import renderer from 'react-test-renderer';

// Queue renders
test("Insertion Sort renders correctly", () => {
    const tree = renderer.create(<InsertionSort/>).toJSON()
    expect(tree).toMatchSnapshot()
});

describe('Testing enqueue', () => {
    const sort = (array) => {
        let a = [...array]
        for (let j = 1; j < a.length; j++) {
          let key = a[j]
          let i = j - 1
          while (i >= 0 && a[i] > key) {
            a[i + 1] = a[i]
            i--
          }
          a[i + 1] = key
        }
        return a 
    }

    it('sorts correctly', ()=>{
      let array = [5,4,3,2,1]
      expect(sort(array)).toStrictEqual([1,2,3,4,5]);
    })

    it('sorts correctly', ()=>{
        let array = [98,67,8,456,3,45,44,4]
        expect(sort(array)).toStrictEqual([3,4,8,44,45,67,98,456]);
    })

    it('sorts correctly', ()=>{
        let array = [1,2,3,4,5,6,7,8,9,10]
        expect(sort(array)).toStrictEqual([1,2,3,4,5,6,7,8,9,10]);
    })
});