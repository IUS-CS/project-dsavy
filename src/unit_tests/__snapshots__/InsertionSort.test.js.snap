// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Insertion Sort renders correctly 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <form
      onSubmit={[Function]}
    >
      <button
        className="my-1 btn btn-primary"
        disabled={false}
        type="submit"
      >
        Insert
      </button>
      <input
        className="enqueue"
        onChange={[Function]}
        title="queue-input"
        type="number"
        value=""
      />
    </form>
  </div>
  <div
    className="row"
  >
    <button
      className="my-1 btn btn-primary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Remove
    </button>
  </div>
  <div
    className="row"
  >
    <button
      className="my-1"
      onClick={[Function]}
    >
      Sort
    </button>
  </div>
  <div
    className="row"
  >
    <div
      style={
        Object {
          "border": "solid",
          "height": "18rem",
          "margin": "0 auto",
          "position": "relative",
          "width": "50rem",
        }
      }
    >
      <ul />
    </div>
  </div>
</div>
`;
